src/main/java/com/google/devtools/build/lib/analysis/actions/SpawnAction.java

```diff
@@ -403,32 +419,70 @@ public Spawn getSpawn(ActionExecutionContext actionExecutionContext)
       throws CommandLineExpansionException, InterruptedException {
     return getSpawn(
         actionExecutionContext.getArtifactExpander(),
-        actionExecutionContext.getClientEnv(),
-        /*envResolved=*/ false,
-        actionExecutionContext.getTopLevelFilesets(),
-        /*reportOutputs=*/ true);
+        actionExecutionContext.getMetadataHandler(),
+        actionExecutionContext.getExecRoot(),
+        /*envResolved=*/ actionExecutionContext.getClientEnv(),
+        false,
+        /*reportOutputs=*/ actionExecutionContext.getTopLevelFilesets(), true);
   }
 
   /**
    * Return a spawn that is representative of the command that this Action will execute in the given
    * environment.
    *
-   * @param envResolved If set to true, the passed environment variables will be used as the Spawn
-   *     effective environment. Otherwise they will be used as client environment to resolve the
-   *     action env.
+   * @param metadataHandler
+   * @param envResolved     If set to true, the passed environment variables will be used as the
+   *                        Spawn effective environment. Otherwise they will be used as client
+   *                        environment to resolve the action env.
    */
   protected Spawn getSpawn(
       ArtifactExpander artifactExpander,
+      MetadataHandler metadataHandler,
+      Path execRoot,
       Map<String, String> env,
       boolean envResolved,
       Map<Artifact, ImmutableList<FilesetOutputSymlink>> filesetMappings,
       boolean reportOutputs)
       throws CommandLineExpansionException, InterruptedException {
+    PathMapper pathMapper;
+    CommandAdjuster pathStripper;
+    if ((outputPathsMode == OutputPathsMode.PER_ACTION
+        || outputPathsMode == OutputPathsMode.CONTENT) && executionInfo.containsKey(
+        ExecutionRequirements.SUPPORTS_PATH_MAPPING)) {
+      try (SilentCloseable c = Profiler.instance().profile("PrecomputedPathMapper.create")) {
+        if (outputPathsMode == OutputPathsMode.PER_ACTION) {
+          pathMapper = PrecomputedPathMapper.createPerAction(artifactExpander, metadataHandler,
+              getInputs(), getPrimaryOutput());
+        } else {
+          pathMapper = PrecomputedPathMapper.createContentBased(artifactExpander, metadataHandler,
+              getInputs(), getPrimaryOutput());
+        }
+      }
+      if (executionInfo.containsKey(ExecutionRequirements.REQUIRES_MATERIALIZED_INPUT_PATHS)) {
+        try {
+          // TODO(fmeum): This is most likely not the right place to materialize the mapped paths:
```

Comment:
This is probably getting advanced, but does this work with virtual inputs, like .params files? See, for example, https://github.com/bazelbuild/bazel/blob/361ce673ad2b959b73e859a121ff2e996feb561b/src/main/java/com/google/devtools/build/lib/exec/ExecutionOptions.java#L123 which defaults to false.

More broadly, actions don't necessarily write outputs to the file system. Would this still try to symlink them?